--- gstiirequalizer.c.orig	2008-06-19 12:45:46.000000000 -0400
+++ gstiirequalizer.c	2008-06-19 12:45:58.000000000 -0400
@@ -27,8 +27,6 @@
 #include <string.h>
 
 #include "gstiirequalizer.h"
-#include "gstiirequalizernbands.h"
-#include "gstiirequalizer3bands.h"
 #include "gstiirequalizer10bands.h"
 
 GST_DEBUG_CATEGORY (equalizer_debug);
@@ -725,15 +723,7 @@
 {
   GST_DEBUG_CATEGORY_INIT (equalizer_debug, "equalizer", 0, "equalizer");
 
-  if (!(gst_element_register (plugin, "equalizer-nbands", GST_RANK_NONE,
-              GST_TYPE_IIR_EQUALIZER_NBANDS)))
-    return FALSE;
-
-  if (!(gst_element_register (plugin, "equalizer-3bands", GST_RANK_NONE,
-              GST_TYPE_IIR_EQUALIZER_3BANDS)))
-    return FALSE;
-
-  if (!(gst_element_register (plugin, "equalizer-10bands", GST_RANK_NONE,
+  if (!(gst_element_register (plugin, "banshee-equalizer", GST_RANK_NONE,
               GST_TYPE_IIR_EQUALIZER_10BANDS)))
     return FALSE;
 
